drivers/hmis/app/graphql/mutations/update_custom_enrollment_value.rb

```diff
@@ -0,0 +1,47 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module Mutations
+  class UpdateCustomEnrollmentValue < BaseMutation
+    argument :enrollment_id, ID, required: true
+    argument :custom_data_element_definition_id, ID, required: true
+    argument :value_float, Float, required: false
+    argument :value_integer, Integer, required: false
+    argument :value_boolean, Boolean, required: false
+    argument :value_string, String, required: false
+    argument :value_text, String, required: false
+    argument :value_date, GraphQL::Types::ISO8601Date, required: false
+
+    field :enrollment, Types::HmisSchema::Enrollment, null: false
+
+    def resolve(enrollment_id:, custom_data_element_definition_id:, **value_kwargs)
+      enrollment = Hmis::Hud::Enrollment.viewable_by(current_user).find_by(id: enrollment_id)
+      raise 'Enrollment not found' unless enrollment.present?
+      raise 'Access denied' unless current_user.permissions_for?(enrollment, :can_edit_enrollments)
+
+      no_value = value_kwargs.compact_blank.empty?
+
+      cded = Hmis::Hud::CustomDataElementDefinition.find_by(id: custom_data_element_definition_id)
+      raise 'Custom element must be tied to Enrollment' unless cded.owner_type == 'Hmis::Hud::Enrollment'
+      raise 'Custom element cannot be updated at occurrence' unless cded.at_occurrence
+      raise 'Updating repeating element at occurrence is not yet supported' if cded.repeats
+
+      custom_value = Hmis::Hud::CustomDataElement.where(
+        data_element_definition: cded,
+        data_source_id: enrollment.data_source_id,
+        owner: enrollment,
+      ).first_or_initialize
```

Comment:
Right now we do the validation in AR because custom data elements can be unique or repeated based on the definition. An example is given here: https://github.com/greenriver/hmis-warehouse/blob/28c5222bbb76fadb393c0b19e4a8ba3e315d8c59/drivers/hmis/app/models/hmis/hud/custom_data_element.rb#L31-L36

Reply:
Thanks for clarifying. I was thinking of a conditional unique index for singleton CDEs but that actually doesn't work since `repeats` is on a different table. I suppose if there was a sequence or sort order on CDEs, that might allow for a unique constraint

The AR validation does still allow for dupes to slip in (albeit in a narrow timing window). If we wanted to address that, maybe we could synchronize around the enrollment?
```ruby
  enrollment.with_lock do
      custom_value = Hmis::Hud::CustomDataElement.where(
        data_element_definition: cded,
        data_source_id: enrollment.data_source_id,
        owner: enrollment,
      ).first_or_initialize
      ...
      save! etc
  ...
```

Could just ignore this issue for now though, maybe we could take a pass at improving mutation synchronization more generally down the road

Reply:
Ah okay, I see. We definitely have other potential race conditions across other mutations. I am going to skip it right now but created a ticket https://www.pivotaltracker.com/story/show/185708758